paths.go

```diff
@@ -0,0 +1,128 @@
+package quamina
+
+import (
+	"fmt"
+	"strings"
+)
+
+const PathSeparator = "\n"
+
+type pathsIndex struct {
+	root bool
+
+	nodes map[string]*pathsIndex
+
+	// fields map from it's name to it's full path (as specific in quamina.Field)
+	// will be present only on the leafs
+	fields map[string][]byte
+}
+
+func newPathsIndex() *pathsIndex {
+	return &pathsIndex{
+		root:   true,
+		nodes:  make(map[string]*pathsIndex),
+		fields: make(map[string][]byte),
+	}
+}
+
+func (p *pathsIndex) add(path string) {
+	parts := strings.Split(path, PathSeparator)
+
+	if len(parts) == 1 {
+		// It's a direct field.
+		p.fields[path] = []byte(path)
+		return
+	}
+
+	var node *pathsIndex
+
+	node = p.getOrCreate(parts[0])
+
+	for i, part := range parts[1:] {
+		if i == len(parts)-2 {
+			node.addField(part, []byte(path))
+		} else {
+			node = node.getOrCreate(part)
+		}
+	}
+}
+
+func (p *pathsIndex) Get(name []byte) (HierarchalNameTracker, bool) {
+	n, ok := p.nodes[string(name)]
+	return n, ok
+}
+
+func (p *pathsIndex) getOrCreate(name string) *pathsIndex {
+	if _, ok := p.nodes[name]; !ok {
+		p.nodes[name] = &pathsIndex{
+			root:   false,
+			nodes:  make(map[string]*pathsIndex),
+			fields: make(map[string][]byte),
+		}
+	}
+
+	return p.nodes[name]
+}
+
+func (p *pathsIndex) addField(name string, path []byte) {
+	if _, ok := p.fields[name]; !ok {
+		p.fields[name] = path
+	}
+}
+
+func (p *pathsIndex) IsRoot() bool {
+	return p.root
+}
+
+func (p *pathsIndex) IsNameUsed(name []byte) bool {
+	_, isField := p.fields[string(name)]
+	if isField {
+		return true
+	}
+	_, isNode := p.nodes[string(name)]
+	return isNode
+}
+
+func (p *pathsIndex) PathForMemberName(name []byte) []byte {
+	return p.fields[string(name)]
+}
+
+func (p *pathsIndex) NodesCount() int {
+	return len(p.nodes)
+}
+
+func (p *pathsIndex) FieldsCount() int {
+	return len(p.fields)
+}
+
+func (p *pathsIndex) String() string {
```

Comment:
is this for debugging/testing or actually called at runtime or in error-handling? If just for debug/test, please have a comment saying so.  BTW I'm totally in favor of having `String()` func for any reasonably complex type to facilitate debugging.

Reply:
It's for debugging, will add a comment for that 👍

Reply:
Added a comment.